enhancements/installer/operators-check.md

```diff
@@ -0,0 +1,174 @@
+  - None
+creation-date: 2021-07-14
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - none
+see-also:
+  - "/enhancements/this-other-neat-thing.md"
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Cluster Operator Installation Check
+
+## Summary
+
+This enhancement proposes that the OpenShift Installer should check the status
+of individual cluster operators and use those statuses as a threshold for 
+determining whether an install is a success (exit 0) or failure (exit != 0).
+Specifically, the Installer should check whether cluster operators have stopped
+progressing for a certain amount of time. If the Installer sees that an operator
+is Available=true but fails to enter a stable Progressing=false state, the Installer
+will exit non-zero and log an error.
+
+## Motivation
+
+This enhancement allows the Installer to identify and catch a class of errors
+which are currently going unchecked. Without this enhancement, the Installer can
+(and does) declare clusters successfully installed when the cluster has failing components.
+Adding a check for cluster operator stability will allow normal users and managed services teams
+to more easily identify and troubleshoot issues when they encounter this class of failure. The change
+will also help the Technical Release Team identify this class of problem in CI and triage
+issues to the appropriate operator development team.
+
+### User Stories
+
+As an openshift-install user, I want the Installer to tell me when an operator never stops
+progressing so that I can check whether the operator has an issue.
+
+As a member of the technical release team, I want the Installer to exit non-zero when
+an operator never stops progressing so that I can triage operator bugs.
+
+### Goals
+
+* Installer correctly identifies a failed cluster install where cluster operators are not stable
+* Installer exits non-zero in those situations and delivers meaningful error message
+
+### Non-Goals
+
+* Installer handling other classes of errors, such as failure to provision machines by MAO
+
+## Proposal
+
+### Workflow Description
+
+Cluster admin will begin installing cluster as usual. The Installation workflow will be:
+
+1. Cluster initializes as usual
+2. As usual, installer checks that cluster version is Available=true Progressing=False Degraded=False 
+3. Installer checks status of each cluster operator for stability
+4. If a cluster operator is not stable the installer logs a message and throws an error
+
+### API Extensions
+
+None
+
+### Implementation Details/Notes/Constraints [optional]
+
+The important detail is how we define _stable_. The definition proposed here is:
+
+```
+If a cluster operator does not maintain Progressing=False for at least 30 seconds,
+during a five minute period it is unstable.
```

Comment:
can we get some more detail on how this will work in practice?

At what point will the installer start the "5 minute window" for example?

And presumably it will be looking for a 30 second period during which *all* operators are progressing=false, meaning you could have a situation where:

1) installer starts watching
2) all operators are progressing=false
3) 25 seconds go by
4) operatorA goes progressing=true
5) some time passes (less than 5 mins total have expired)
6) operator A goes progressing=false
7) 25 seconds go by
8) operator B goes progressing=true
9) some time passes (less than 5 mins total have expired)
10) operator B goes progressing=false
11) 25 seconds go by
12) operatorA (or C) goes progressing=true
etc.

right?

Not saying that's a problem, just trying to understand how the detection/determination will work in practice.

Reply:
Do we trust lastTransition Time?

1. Wait for current install-complete signal
2. Start's a five minute timer
3. Every 5s check each ClusterOperators for Progressing=False and lastTransitionTime < now - 30s? 
4. Exit non-zero at end of five minute timer

Reply:
The implementation is here: 
https://github.com/openshift/installer/pull/6124/files#diff-56276d5381d618d46ec8d35d93210662c8fdd4c9bcd90fd36afbc6a59227eb0bR713-R738

The intent It is pretty much what scott said except we don't necessarily wait the whole 5 minutes. `lastTransitionTime < now - 30s` is the level for stability, once that is cleared the CO is cleared. If all COs are "stable" on the first check we exit immediately and the whole check took seconds at most.

The 5s detail in point 3 is abstracted away into library-go & watch events.

Reply:
Scott's explanation in https://github.com/openshift/enhancements/pull/1189/files#r921528565 makes sense to me.  I think the logic in the PR may have a bug.  I commented.